### PR TITLE
change some Exceptions to CatchableError or Defect, fixes #10288

### LINKED ISSUE #10288: Exception types in the stdlib should inherit from `CatchableError` or `Defect`, not `Exception`
Exception types should always inherit from `CatchableError` or `Defect` (directly or indirectly) to indicate if they should be catched or not. Here is a list of exception types in the stdlib that breaks this rule:

- [ ] `nre.RegexError`
- [ ] `asyncfutures.FutureError`
- [ ] `net.SslError`
- [ ] `net.TimeoutError`
- [ ] `punycode.PunyError`
- [ ] `selectors.IOSelectorsException`

### PR DESCRIPTION
**Pull request by metagn**
fixes #10288

### CODE REVIEW

File: lib/impure/nre.nim [hunk 1]
```diff
@@ -198,7 +198,7 @@ type
   Captures* = distinct RegexMatch
   CaptureBounds* = distinct RegexMatch
 
-  RegexError* = ref object of Exception
+  RegexError* = ref object of CatchableError
 
   RegexInternalError* = ref object of RegexError
```
Comment (hlaaftana) [inline lib/impure/nre.nim:203]: I debated making RegexInternalError extend Defect instead of RegexError but didn't do it, should I do it?

Comment (metagn) [inline lib/impure/nre.nim:203]: I debated making RegexInternalError extend Defect instead of RegexError but didn't do it, should I do it?

**Comment by arnetheduck** [comment]
1.2 backport?

File: lib/impure/nre.nim (hunk 1 above)
Comment (Araq) [reply lib/impure/nre.nim:203]: RegexErrors are ValueErrors

**Pull request merged**